state/protocol/protocol_state/kvstore/models.go

```diff
@@ -190,9 +208,55 @@ func (model *Modelv1) GetProtocolStateVersion() uint64 {
 	return 1
 }
 
+// Modelv2 is v2 of the Protocol State key-value store.
+// This version adds the following changes:
+//   - Non-system-chunk service event validation support (adds ChunkBody.ServiceEventCount field)
+//   - EFM Recovery (adds EpochCommit.DKGIndexMap field)
+type Modelv2 struct {
+	Modelv1
+}
+
+// ID returns an identifier for this key-value store snapshot by hashing internal fields and version number.
+func (model *Modelv2) ID() flow.Identifier {
+	return makeVersionedModelID(model)
+}
+
+// Replicate instantiates a Protocol State Snapshot of the given protocolVersion.
+// It clones existing snapshot if protocolVersion = currentVersion, other versions are not supported yet.
+// Expected errors during normal operations:
+//   - ErrIncompatibleVersionChange if replicating the Parent Snapshot into a Snapshot
+//     with the specified `protocolVersion` is not supported.
+func (model *Modelv2) Replicate(protocolVersion uint64) (protocol_state.KVStoreMutator, error) {
+	currentVersion := model.GetProtocolStateVersion()
+	if currentVersion == protocolVersion {
+		// no need for migration, return a complete copy
+		return clone.Clone(model), nil
+	} else {
+		return nil, fmt.Errorf("unsupported replication currentVersion %d: %w",
+			protocolVersion, ErrIncompatibleVersionChange)
+	}
+}
+
+// VersionedEncode encodes the key-value store, returning the version separately
+// from the encoded bytes.
+// No errors are expected during normal operation.
+func (model *Modelv2) VersionedEncode() (uint64, []byte, error) {
+	return versionedEncode(model.GetProtocolStateVersion(), model)
+}
+
+// GetProtocolStateVersion returns the version of the Protocol State Snapshot
+// that is backing the `Reader` interface. It is the protocol version that originally
+// created the Protocol State Snapshot. Changes in the protocol state version
+// correspond to changes in the set of key-value pairs which are supported,
+// and which model is used for serialization.
+func (model *Modelv2) GetProtocolStateVersion() uint64 {
+	return 2
+}
+
 // NewDefaultKVStore constructs a default Key-Value Store of the *latest* protocol version for bootstrapping.
 // Currently, the KV store is largely empty.
 // TODO: Shortcut in bootstrapping; we will probably have to start with a non-empty KV store in the future;
+// TODO(efm-recovery): we need to bootstrap with v1 in order to test the upgrade to v2. Afterward, we should bootstrap with v2 by default for new networks.
```

Comment:
This function is only used for network-scale bootstrapping, hence why the comment is talking about networks rather than nodes. A new node would bootstrap with whatever version the network is currently running, based on the root protocol state snapshot it boots from.